pkg/controller/direct/iam/iambindings.go

```diff
@@ -0,0 +1,252 @@
+// from all the existing bindings from the underlying IAM Policy.
+// The status.LastAppliedBindings in the returned IAMPartialPolicy will be cleared.
+func ComputePartialPolicyWithRemainingBindings(partialPolicy *v1beta1.IAMPartialPolicy, livePolicy *v1beta1.IAMPolicy) *v1beta1.IAMPartialPolicy {
+	desiredPartialPolicy := partialPolicy.DeepCopy()
+	remainingBindings := removeMembersPerBinding(livePolicy.Spec.Bindings, partialPolicy.Status.LastAppliedBindings)
+	// record the remaining bindings as (new) all bindings.
+	sortBindingSlice(remainingBindings)
+	desiredPartialPolicy.Status.AllBindings = remainingBindings
+	// clear last applied bindings
+	desiredPartialPolicy.Status.LastAppliedBindings = make([]v1beta1.IAMPolicyBinding, 0)
+	return desiredPartialPolicy
+}
+
+func ConvertIAMPartialBindingsToIAMPolicyBindings(partialPolicy *v1beta1.IAMPartialPolicy, resolver MemberIdentityResolver) (bindings []v1beta1.IAMPolicyBinding, err error) {
+	res := make([]v1beta1.IAMPolicyBinding, 0)
+	for _, binding := range partialPolicy.Spec.Bindings {
+		convertedBinding, err := toIAMPolicyBinding(binding, resolver, partialPolicy.Namespace)
+		if err != nil {
+			return bindings, fmt.Errorf("error converting IAMPartialPolicy binding to IAMPolicy binding: %w", err)
+		}
+		res = append(res, convertedBinding)
+	}
+	return mergeBindingsWithSameRoleAndCondition(res), nil
+}
+
+func toIAMPolicyBinding(b v1beta1.IAMPartialPolicyBinding, resolver MemberIdentityResolver, defaultNamespace string) (binding v1beta1.IAMPolicyBinding, err error) {
+	members := make([]v1beta1.Member, 0)
+	for _, m := range b.Members {
+		resolvedMember, err := resolver.Resolve(m.Member, m.MemberFrom, defaultNamespace)
+		if err != nil {
+			return binding, fmt.Errorf("error resolving member identity of IAMPartialPolicy binding: %w", err)
+		}
+		members = append(members, v1beta1.Member(resolvedMember))
+	}
+
+	return v1beta1.IAMPolicyBinding{
+		Role:      b.Role,
+		Condition: b.Condition,
+		Members:   members,
+	}, nil
+}
+
+func mergeBindingsWithSameRoleAndCondition(bindings []v1beta1.IAMPolicyBinding) []v1beta1.IAMPolicyBinding {
+	bindingMap := mergeBindings(bindings)
+	mergedBindings := make([]v1beta1.IAMPolicyBinding, 0)
+	for _, v := range bindingMap {
+		if len(v.Members) > 0 {
+			mergedBindings = append(mergedBindings, v)
+		}
+	}
+	return mergedBindings
+}
+
+func mergeBindingSlices(bindingSlice1, bindingSlice2 []v1beta1.IAMPolicyBinding) []v1beta1.IAMPolicyBinding {
+	mergedBindings := make([]v1beta1.IAMPolicyBinding, 0)
+	mergedBindings = append(mergedBindings, bindingSlice1...)
+	mergedBindings = append(mergedBindings, bindingSlice2...)
+	return mergeBindingsWithSameRoleAndCondition(mergedBindings)
+}
+
+func mergeBindings(bindings []v1beta1.IAMPolicyBinding) map[iamBindingKey]v1beta1.IAMPolicyBinding {
+	bindingMap := make(map[iamBindingKey]v1beta1.IAMPolicyBinding)
+	for _, a := range bindings {
```

Comment:
Not asking for long variable names but something better than a single letter. 'binding' seems better than 'a', 'key' seems better than 'k'.